miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultInstrumentService.java

```diff
@@ -71,9 +77,74 @@ public long create(Instrument instrument) throws IOException {
     return save(instrument);
   }
 
+  @Override
+  public void addServiceRecord(ServiceRecord record, Instrument instrument) throws IOException {
+    // originally located at HibernateServiceRecordDao
+    if (instrument.getDateCommissioned() != null) {
+      throw new IOException("Cannot add service records to a retired instrument!");
+    }
+    long recordId = serviceRecordService.create(record);
+    ServiceRecord serviceRecord = serviceRecordService.get(recordId);
+
+    // add saved service record
+    instrument.getServiceRecords().add(serviceRecord);
+
+    // save instrument
+    save(instrument);
+  }
+
+  @Override
+  public void removeServiceRecord(ServiceRecord record, Instrument instrument) throws IOException {
+    if (!instrument.getServiceRecords().contains(record)) {
+      throw new IOException("Cannot remove service record that does not exist");
+    }
+    instrument.getServiceRecords().remove(record);
+    save(instrument);
+
+    // remove servicerecord
+    serviceRecordService.delete(record);
+  }
+
+  @Override
+  public void updateServiceRecord(ServiceRecord record, Instrument instrument) throws IOException {
+
+    if (!instrument.getServiceRecords().contains(record)) {
+      throw new IOException("Cannot update service record that does not exist");
+    }
+    long recordId = record.getId();
+    InstrumentPosition position = findPosition(recordId, instrument);
+    // remove the old service record from the instrument
+    instrument.getServiceRecords().remove(record);
+
+    // update the servicerecord
+    // recordId = serviceRecordService.update(record, position);
+    recordId = serviceRecordService.update(record);
+    ServiceRecord serviceRecord = serviceRecordService.get(recordId);
+
+    // add new service record to the instrument
+    instrument.getServiceRecords().add(serviceRecord);
+
+    // update instrument
+    // update(instrument, position);
+
+    // save
+    save(instrument);
+  }
+
+
+  private InstrumentPosition findPosition(long id, Instrument instrument) {
+    // must ask Dillan about how this should be handled.
+    // some functions that require InstrumentPosition are commented out for this reason
+    return instrument.getInstrumentModel().getPositions().stream()
+        .filter(p -> p.getId() == id)
+        .findFirst().orElse(null);
+  }
+
   @Override
   public long update(Instrument instrument) throws IOException {
-    authorizationManager.throwIfNonAdmin();
+    if (instrument == null) {
+      throw new IOException("Cannot update instrument that does not exist");
+    }
```

Comment:
What was your thought behind including this check?

Reply:
I thought I would need to check if instrument is null when updating the service record, but now that I think of it, I don't think this is necessary because we need a check on managedInstrument. I will delete this.